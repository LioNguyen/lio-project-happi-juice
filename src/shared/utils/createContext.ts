import type React from "react";
import { createContext as originCreateContext, useContext } from "react";

function createContext<T>(): [React.Provider<T>, () => T, React.Context<T>] {
  const context = originCreateContext<T>({} as T);

  return [context.Provider, () => useContext(context), context];
}

export { createContext };
