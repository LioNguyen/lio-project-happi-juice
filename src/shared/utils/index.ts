export * from './lazyLoad'
export * from './tailwind'
