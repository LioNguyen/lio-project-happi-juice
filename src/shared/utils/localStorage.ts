/**
 * Save data to localStorage
 * @param key - The key under which the data will be stored
 * @param value - The data to store, can be any serializable value
 */
function setLocalStorage<T>(key: string, value: T): void {
  try {
    const serializedValue =
      typeof value === 'string' ? value : JSON.stringify(value)

    localStorage.setItem(key, serializedValue)
  } catch (error) {
    console.error(`Error saving ${key} to localStorage:`, error)
  }
}

/**
 * Retrieve data from localStorage
 * @param key - The key of the stored data
 * @returns The parsed data or null if not found or error occurs
 */
function getLocalStorage<T>(key: string): T | null {
  try {
    const serializedValue = localStorage.getItem(key)

    if (!serializedValue || serializedValue === 'undefined') {
      return null
    }

    // Try to parse JSON
    try {
      return JSON.parse(serializedValue) as T
    } catch {
      // Return as string if JSON parsing fails
      return serializedValue as T
    }
  } catch (error) {
    console.error(`Error reading ${key} from localStorage:`, error)
    return null
  }
}

/**
 * Remove data from localStorage
 * @param key - The key to remove from localStorage
 */
function removeLocalStorage(key: string): void {
  try {
    localStorage.removeItem(key)
  } catch (error) {
    console.error(`Error removing ${key} from localStorage:`, error)
  }
}

/**
 * Clear all data from localStorage
 */
function clearLocalStorage(): void {
  try {
    localStorage.clear()
  } catch (error) {
    console.error('Error clearing localStorage:', error)
  }
}

export {
  clearLocalStorage,
  getLocalStorage,
  removeLocalStorage,
  setLocalStorage,
}
