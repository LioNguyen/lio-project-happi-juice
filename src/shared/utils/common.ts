import { format, isValid, Locale, parseISO } from 'date-fns'
import { enUS, vi } from 'date-fns/locale'
import _ from 'lodash'

import { IOrderItem } from '@/domains/order'
import { ISelectOption } from '@/shared/types'

interface INumberFormatOptions {
  decimals?: number
  thousandSeparator?: string
  decimalSeparator?: string
}

// ==================== Date Utilities ====================
/**
 * Parse various date formats into Date object
 * @category Date
 * @param {Date | string | undefined} value - Input date
 * @returns {Date | undefined} Parsed Date object or undefined if invalid
 * @example
 * parseDate("2024-02-26") // Date object
 * parseDate("invalid") // undefined
 */
function parseDate(value: Date | string | undefined): Date | undefined {
  if (!value) return undefined
  if (value instanceof Date) return isValid(value) ? value : undefined

  try {
    // Try ISO string parsing first
    const date = parseISO(value)
    if (isValid(date)) return date

    // Fallback to direct Date creation
    const normalDate = new Date(value)
    return isValid(normalDate) ? normalDate : undefined
  } catch {
    return undefined
  }
}

/**
 * Format a date string according to specified format
 * @category Date
 * @param {string} date - Date string to format
 * @param {string} [dateFormat] - Output format (defaults to "yyyy-MM-dd")
 * @returns {string} Formatted date string
 * @example
 * formatDateTime("2024-02-26") // "2024-02-26"
 * formatDateTime("2024-02-26", "dd/MM/yyyy") // "26/02/2024"
 */
function formatDateTime(date: string, dateFormat?: string) {
  return format(date, dateFormat || 'yyyy-MM-dd')
}

const localeMap: Record<string, Locale> = {
  vi: vi,
  en: enUS,
}

// ==================== Number & Price Utilities ====================
/**
 * Format a number with customizable separators and decimals
 * @category Number
 * @param {number} value - Number to format
 * @param {INumberFormatOptions} [options] - Formatting options
 * @returns {string} Formatted number string
 */
const formatNumber = (
  value: number,
  options?: INumberFormatOptions,
): string => {
  const {
    decimals = 0,
    thousandSeparator = ',',
    decimalSeparator = '.',
  } = options || {}

  const parts = value.toFixed(decimals).split('.')
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, thousandSeparator)

  return parts.join(decimalSeparator)
}

/**
 * Formats the price with thousand separators using locale-specific formatting
 * @category Price
 * @param {string} price - Price string to format
 * @returns {string} Formatted price string
 */
const formatPrice = (price: string) => {
  return Number(price).toLocaleString()
}

// ==================== Array Utilities ====================
/**
 * Transforms an array of select options into a sorted, joined string
 * @category Array
 * @param {ISelectOption[]} array - Array of select options
 * @param {string} [key] - Key to extract from each option
 * @param {string} [separator] - Separator for joining values
 * @returns {string} Sorted and joined string
 */
const transformToSortedString = (
  array: ISelectOption[],
  key: string = 'value',
  separator: string = ',',
): string => {
  return _.chain(array).map(key).sortBy().join(separator).value()
}

// ==================== Order Utilities ====================
/**
 * Calculates the total price for an array of order items
 * @category Order
 * @param {IOrderItem[]} items - Array of order items
 * @returns {number} Total price
 */
const calculateTotalPrice = (items: IOrderItem[]) => {
  return items.reduce((sum, item) => sum + item.price * item.quantity, 0)
}

export {
  // Date utils
  parseDate,
  formatDateTime,
  localeMap,

  // Number & Price utils
  formatNumber,
  formatPrice,

  // Array utils
  transformToSortedString,
  // Order utils
  calculateTotalPrice,
}
