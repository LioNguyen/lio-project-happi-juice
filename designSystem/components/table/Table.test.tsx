import { screen } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import React from "react";

import { renderWithTheme } from "@/shared/utils";
import { Table, TableHeader, TableBody, TableFooter, TableHead, TableRow, TableCell, TableCaption } from "./Table";

describe("Table Component", () => {
  const renderTable = (props = {}) => {
    return renderWithTheme(
      <Table {...props}>
        <TableCaption>A list of your recent invoices.</TableCaption>
        <TableHeader>
          <TableRow>
            <TableHead>Invoice</TableHead>
            <TableHead>Status</TableHead>
            <TableHead>Method</TableHead>
            <TableHead>Amount</TableHead>
          </TableRow>
        </TableHeader>
        <TableBody>
          <TableRow>
            <TableCell>INV001</TableCell>
            <TableCell>Paid</TableCell>
            <TableCell>Credit Card</TableCell>
            <TableCell>$250.00</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>INV002</TableCell>
            <TableCell>Pending</TableCell>
            <TableCell>PayPal</TableCell>
            <TableCell>$150.00</TableCell>
          </TableRow>
        </TableBody>
        <TableFooter>
          <TableRow>
            <TableCell colSpan={3}>Total</TableCell>
            <TableCell>$400.00</TableCell>
          </TableRow>
        </TableFooter>
      </Table>,
    );
  };

  // Basic Rendering Tests
  it("should render all table components", () => {
    renderTable();

    expect(screen.getByRole("table")).toBeInTheDocument();
    expect(screen.getByRole("caption")).toBeInTheDocument();
    expect(screen.getAllByRole("rowgroup")).toHaveLength(3); // header, body, footer
    expect(screen.getAllByRole("row")).toHaveLength(4); // 1 header + 2 body + 1 footer
    expect(screen.getAllByRole("cell")).toHaveLength(10); // (2 rows × 4 cells) + (1 footer row × 2 cells)
    expect(screen.getAllByRole("columnheader")).toHaveLength(4);
  });

  // Snapshot Tests
  describe("Table Snapshots", () => {
    it("should match snapshot with default props", () => {
      const { container } = renderTable();
      expect(container).toMatchSnapshot();
    });

    it("should match snapshot with custom className", () => {
      const { container } = renderTable({ className: "custom-table" });
      expect(container).toMatchSnapshot();
    });
  });

  // Table Container Tests
  describe("Table Container", () => {
    it("should render with correct default classes", () => {
      renderTable();
      const tableContainer = screen.getByRole("table").parentElement;
      expect(tableContainer).toHaveClass("relative", "w-full", "overflow-auto");
    });
  });

  // Table Tests
  describe("Table", () => {
    it("should render with correct default classes", () => {
      renderTable();
      const table = screen.getByRole("table");
      expect(table).toHaveClass("w-full", "caption-bottom", "text-sm");
    });

    it("should merge custom className with default classes", () => {
      renderTable({ className: "custom-table" });
      const table = screen.getByRole("table");
      expect(table).toHaveClass("w-full", "caption-bottom", "text-sm", "custom-table");
    });
  });

  // TableHeader Tests
  describe("TableHeader", () => {
    it("should render with correct default classes", () => {
      renderTable();
      const headers = screen.getAllByRole("rowgroup");
      expect(headers[0]).toHaveClass("[&_tr]:border-b");
    });

    it("should merge custom className", () => {
      renderWithTheme(
        <Table>
          <TableHeader className="custom-header">
            <TableRow>
              <TableHead>Header</TableHead>
            </TableRow>
          </TableHeader>
        </Table>,
      );
      const header = screen.getByRole("rowgroup");
      expect(header).toHaveClass("[&_tr]:border-b", "custom-header");
    });
  });

  // TableBody Tests
  describe("TableBody", () => {
    it("should render with correct default classes", () => {
      renderTable();
      const bodies = screen.getAllByRole("rowgroup");
      const body = bodies[1]; // Second rowgroup is TableBody
      expect(body).toHaveClass("[&_tr:last-child]:border-0");
    });
  });

  // TableFooter Tests
  describe("TableFooter", () => {
    it("should render with correct default classes", () => {
      renderTable();

      // Get last rowgroup which is footer
      const footer = screen.getAllByRole("rowgroup")[2];

      // Test individual classes
      expect(footer).toHaveClass("border-t");
      expect(footer).toHaveClass("bg-muted/50");
      expect(footer).toHaveClass("font-medium");
      expect(footer).toHaveClass("[&>tr]:last:border-b-0");
    });
  });

  // TableRow Tests
  describe("TableRow", () => {
    it("should render with correct default classes", () => {
      renderTable();
      const rows = screen.getAllByRole("row");
      expect(rows[1]).toHaveClass(
        "border-b",
        "transition-colors",
        "hover:bg-muted/50",
        "data-[state=selected]:bg-muted",
      );
    });

    it("should render selected state correctly", () => {
      renderWithTheme(
        <Table>
          <TableBody>
            <TableRow data-state="selected">
              <TableCell>Selected Row</TableCell>
            </TableRow>
          </TableBody>
        </Table>,
      );
      const row = screen.getByRole("row");
      expect(row).toHaveAttribute("data-state", "selected");
    });
  });

  // TableHead Tests
  describe("TableHead", () => {
    it("should render with correct default classes", () => {
      renderTable();
      const headers = screen.getAllByRole("columnheader");
      headers.forEach(header => {
        expect(header).toHaveClass("h-10", "px-2", "text-left", "align-middle", "font-medium", "text-muted-foreground");
      });
    });

    it("should handle checkbox styling correctly", () => {
      renderWithTheme(
        <Table>
          <TableHeader>
            <TableRow>
              <TableHead>
                <input type="checkbox" role="checkbox" />
              </TableHead>
            </TableRow>
          </TableHeader>
        </Table>,
      );
      const header = screen.getByRole("columnheader");
      expect(header).toHaveClass("[&:has([role=checkbox])]:pr-0");
    });
  });

  // TableCell Tests
  describe("TableCell", () => {
    it("should render with correct default classes", () => {
      renderTable();
      const cells = screen.getAllByRole("cell");
      cells.forEach(cell => {
        expect(cell).toHaveClass("p-2", "align-middle");
      });
    });

    it("should handle checkbox styling correctly", () => {
      renderWithTheme(
        <Table>
          <TableBody>
            <TableRow>
              <TableCell>
                <input type="checkbox" role="checkbox" />
              </TableCell>
            </TableRow>
          </TableBody>
        </Table>,
      );
      const cell = screen.getByRole("cell");
      expect(cell).toHaveClass("[&:has([role=checkbox])]:pr-0");
    });
  });

  // TableCaption Tests
  describe("TableCaption", () => {
    it("should render with correct default classes", () => {
      renderTable();
      const caption = screen.getByRole("caption");
      expect(caption).toHaveClass("mt-4", "text-sm", "text-muted-foreground");
    });

    it("should merge custom className", () => {
      renderWithTheme(
        <Table>
          <TableCaption className="custom-caption">Caption Text</TableCaption>
        </Table>,
      );
      const caption = screen.getByRole("caption");
      expect(caption).toHaveClass("mt-4", "text-sm", "text-muted-foreground", "custom-caption");
    });
  });

  // Ref Forwarding Tests
  describe("Ref Forwarding", () => {
    it("should forward ref to table element", () => {
      const ref = React.createRef<HTMLTableElement>();
      renderWithTheme(<Table ref={ref} />);
      expect(ref.current).toBeInstanceOf(HTMLTableElement);
    });

    it("should forward ref to header element", () => {
      const ref = React.createRef<HTMLTableSectionElement>();
      renderWithTheme(<TableHeader ref={ref} />);
      expect(ref.current).toBeInstanceOf(HTMLTableSectionElement);
    });

    // Add similar tests for other components that forward refs
  });
});
