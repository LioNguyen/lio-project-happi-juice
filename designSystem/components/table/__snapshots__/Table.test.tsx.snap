// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table Component > Table Snapshots > should match snapshot with custom className 1`] = `
<div>
  <div
    class="relative w-full overflow-auto"
  >
    <table
      class="w-full caption-bottom text-sm custom-table"
    >
      <caption
        class="mt-4 text-sm text-muted-foreground"
      >
        A list of your recent invoices.
      </caption>
      <thead
        class="[&_tr]:border-b"
      >
        <tr
          class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
        >
          <th
            class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            Invoice
          </th>
          <th
            class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            Status
          </th>
          <th
            class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            Method
          </th>
          <th
            class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            Amount
          </th>
        </tr>
      </thead>
      <tbody
        class="[&_tr:last-child]:border-0"
      >
        <tr
          class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
        >
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            INV001
          </td>
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            Paid
          </td>
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            Credit Card
          </td>
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            $250.00
          </td>
        </tr>
        <tr
          class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
        >
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            INV002
          </td>
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            Pending
          </td>
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            PayPal
          </td>
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            $150.00
          </td>
        </tr>
      </tbody>
      <tfoot
        class="border-t bg-muted/50 font-medium [&>tr]:last:border-b-0"
      >
        <tr
          class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
        >
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            colspan="3"
          >
            Total
          </td>
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            $400.00
          </td>
        </tr>
      </tfoot>
    </table>
  </div>
</div>
`;

exports[`Table Component > Table Snapshots > should match snapshot with default props 1`] = `
<div>
  <div
    class="relative w-full overflow-auto"
  >
    <table
      class="w-full caption-bottom text-sm"
    >
      <caption
        class="mt-4 text-sm text-muted-foreground"
      >
        A list of your recent invoices.
      </caption>
      <thead
        class="[&_tr]:border-b"
      >
        <tr
          class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
        >
          <th
            class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            Invoice
          </th>
          <th
            class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            Status
          </th>
          <th
            class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            Method
          </th>
          <th
            class="h-10 px-2 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            Amount
          </th>
        </tr>
      </thead>
      <tbody
        class="[&_tr:last-child]:border-0"
      >
        <tr
          class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
        >
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            INV001
          </td>
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            Paid
          </td>
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            Credit Card
          </td>
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            $250.00
          </td>
        </tr>
        <tr
          class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
        >
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            INV002
          </td>
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            Pending
          </td>
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            PayPal
          </td>
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            $150.00
          </td>
        </tr>
      </tbody>
      <tfoot
        class="border-t bg-muted/50 font-medium [&>tr]:last:border-b-0"
      >
        <tr
          class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
        >
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            colspan="3"
          >
            Total
          </td>
          <td
            class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
          >
            $400.00
          </td>
        </tr>
      </tfoot>
    </table>
  </div>
</div>
`;
